requestStorageAccessForOrigin.bs

```diff
@@ -0,0 +1,203 @@
+
+1. Let |p| be [=a new promise=].
+1. If |doc| is not [=Document/fully active=], then [=reject=] |p| with an "{{InvalidStateError}}" {{DOMException}} and return |p|.
+1. If |doc|'s [=Document/browsing context=] is not a [=top-level browsing context=], [=reject=] |p| with an "{{InvalidStateError}}" {{DOMException}} and return |p|.
+1. If |doc|'s [=Document/origin=] is an [=opaque origin=], [=reject=] |p| with an "{{InvalidStateError}}" {{DOMException}} and return |p|.
+1. Let |parsedURL| be the the result of running the [=URL parser=] on |requestedOrigin|.
+1. If |parsedURL| is failure, [=reject=] |p| with a "{{TypeError}}" {{DOMException}} and return |p|.
+1. Let |origin| be |parsedURL|'s [=origin=].
+1. If |origin| is an [=opaque origin=], [=reject=] |p| with an "{{InvalidStateError}}" {{DOMException}} and return |p|.
+1. If |doc|'s [=Document/origin=] is [=same origin=] with |origin|, [=/resolve=] and return |p|.
+1. Let |descriptor| be a newly created {{TopLevelStorageAccessPermissionDescriptor}} with <code>name</code> set to "<a permission><code>top-level-storage-access</code></a>" and with <code>requestedOrigin</code> set to |origin|.
+1. Let |global| be |doc|'s [=relevant global object=].
+1. Run these steps [=in parallel=]:
+    1. Let |hasAccess| be [=a new promise=].
+    1. [=Determine the top-level storage access policy=] with |descriptor|, |doc| and |hasAccess|.
+    1. [=Queue a global task=] on the [=permission task source=] given |global| to resolve or reject |p| based on the result of |hasAccess|.
+1. Return |p|.
+
+<h3 id="ua-policies">User Agent top-level storage access policies</h3>
+
+To <dfn type="abstract-op">determine if a request has top-level storage access</dfn> with [=request=] |request|, run these steps:
+
+ISSUE: should all cross-site subresource requests behave this way, or only those initiated by the top-level document, or top-level plus those requested from an iframe that is same-site with the top-level document, or only those that have explicitly requested access (in which case a set of requested origins would be maintained)?
+
+<!-- this long chain pulled from fetch, but may be unusual or unsafe -->
+1. Let |settings| be |request|'s [=request/client=]'s [=relevant global object=]'s [=associated document=]'s [=relevant settings object=].
+1. Let |embedded origin| be |request|'s [=request/url=].
+1. Let |descriptor| be a newly created {{TopLevelStorageAccessPermissionDescriptor}} with <code>name</code> set to "<a permission><code>top-level-storage-access</code></a>" and with <code>requestedOrigin</code> set to |embedded origin|.
+1. Let |existing state| be |descriptor|'s [=permission state=] with |settings|.
+1. If |existing state| is [=permission/granted=], return true.
+1. Return false.
+
+To <dfn type="abstract-op">determine the top-level storage access policy</dfn> for {{TopLevelStorageAccessPermissionDescriptor}} |descriptor|, with {{Document}} |doc| and {{Promise}} |p|, run these steps:
```

Comment:
It would be good to know where this is supposed to be running? On the "main thread", or "in parallel"? It seems to kind of interact with main thread state, but earlier it seems to be called in parallel.

Reply:
(removed the `global` variable and the top-level site in favor of plumbing it in; attempted to follow [the docs](https://html.spec.whatwg.org/multipage/infrastructure.html#in-parallel), though I suspect user activation will also have to be plumbed in)